### PR TITLE
libxslt: build & install xsltproc utility

### PR DESCRIPTION
**Pull request by SpaceIm**
address https://github.com/conan-io/conan-center-index/pull/8695#pullrequestreview-957081175

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by puetzk** [comment]
Thanks! Digging a little more it this morning, I see that libxml2 has an `option.include_utils` in #2274 (default `include_utils=True` and CCI only has binaries with that default). https://github.com/conan-io/conan-center-index/pull/2274#issuecomment-679252099 seemed to cite iOS/wasm as a place where one would perhaps want to skip even building utils? I don't see any other recipes in CCI following that pattern, though, and this change to libxslt was 1.5 years later, so I doubt it was actually related.

This change (just unconditionally bring back xsltproc, reverting 12571daec228681e88f7a3c645e9151a3248f866) seems like the simpler thing, but I wouldn't following the libxml2 pattern would be fine with me too if there was a reason to sometimes want it gone.

**Comment by SpaceIm** [comment]
@puetzk It's fine for iOS, xsltproc can be built (I've tested). And anyway xsltproc was built even before this PR, it was removed from the package, that's all.

**Comment by SpaceIm** [comment]
Ok I remember, it requires more patch to build shared (without static) libxslt & xsltproc with msvc:

```
	link.exe /nologo /LIBPATH:bin.msvc /LIBPATH:C:\J\w\prod\BuildSingleReference@2\.conan\data\libxslt\1.1.34\_\_\package\5f2595b729d2d3b81e5355b544a1dbc8df82b4fd\lib /OUT:bin.msvc\xsltproc.exe libxslt_a.lib libexslt_a.lib  libxml2_a.lib ws2_32.lib wsock32.lib int.utils.msvc\xsltproc.obj
LINK : fatal error LNK1181: cannot open input file 'libxslt_a.lib'
```

**Comment by SpaceIm** [comment]
Ok I remember, it requires more patch to build shared (without static) libxslt & xsltproc with msvc:

```
	link.exe /nologo /LIBPATH:bin.msvc /LIBPATH:C:\J\w\prod\BuildSingleReference@2\.conan\data\libxslt\1.1.34\_\_\package\5f2595b729d2d3b81e5355b544a1dbc8df82b4fd\lib /OUT:bin.msvc\xsltproc.exe libxslt_a.lib libexslt_a.lib  libxml2_a.lib ws2_32.lib wsock32.lib int.utils.msvc\xsltproc.obj
LINK : fatal error LNK1181: cannot open input file 'libxslt_a.lib'
```

**Review by madebr — DISMISSED** [review]
Would a test running `xsltproc --help` (or something similar) be useful?

**Review by madebr — APPROVED** [review]

File: recipes/libxslt/all/patches/0001-allow-xsltproc-link-shared-mt.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+--- a/win32/configure.js
++++ b/win32/configure.js
+@@ -394,7 +394,7 @@ if (error != 0) {
+ // with /MT and friends, then we need to enable static linking for xsltproc
+ if (cruntime == "/MT" || cruntime == "/MTd" ||
+ 		cruntime == "/ML" || cruntime == "/MLd") {
+-	buildStatic = 1;
++	// buildStatic = 1;
```
[the remark this answers is not in the document]
Comment (puetzk) [reply recipes/libxslt/all/patches/0001-allow-xsltproc-link-shared-mt.patch:8]: I did build a configuration with things mixed up `-s libxslt:compiler.runtime=MT -o libxslt:shared=True -o libxml2:iconv=False -o libxml2:shared=True libxslt:compiler.runtime=MD` (just to make sure I was getting a libcmt.lib and ucrtbase.dll mix) and, at least in Windows 10 and VS2022, these seemed to be compatible enough in their definition of `FILE*` that xsltproc telling libxml2 to print to stdout worked despite the mixed CRT implementations. So it's undefined behavior in principle, but not obviously broken in current practice.

[the remark this answers is not in the document]
Comment (SSE4) [reply recipes/libxslt/all/patches/0001-allow-xsltproc-link-shared-mt.patch:8]: @puetzk you're absolutely right. DLL + MT is a subject for UB due to different heap.
see also #10349 & #8373
we're thinking about deactivating this configuratiion globally on our CI, as it gives more troubles than benefits.
(another problem was in #7691).
I'll let you know about our decision.
as for `FILE*` and other CRT object, it's definetely different implementation between UCRT and MSVCRT (see also https://www.msys2.org/docs/environments/).
that's one of the reasons why even pure C libraries cannot be safely mixed across different compilers and different compiler version (e.g. MinGW-compiled objects with MSVC-compiled objects).
unless special treatment is done (e.g. all the allocations and deallocations happens inside the same DLL via exported function, and never across DLL boundaties).

[the remark this answers is not in the document]
Comment (puetzk) [reply recipes/libxslt/all/patches/0001-allow-xsltproc-link-shared-mt.patch:8]: Right. Heap seems to actually be accounted for in this case (xsltproc seems to exclusively use xmlMalloc and xmlFree, and thus shares its allocator with libxml2.dll regardless). Even with built with `/MD`, it does not link to api-ms-win-crt-heap.

And my quick testing testing (expecting to provoke a crash as an example of why this configuration is problematic) instead showed that, at least for the VS2022 CRT versions, `FILE *` surprisingly *is* compatible between ucrtbase.dll and libcmt.lib (at least enough so for `stdout` to work when [xsltproc calls `xsltSaveResultToFile(stdout, res, cur)`](https://gitlab.gnome.org/GNOME/libxslt/-/blob/master/xsltproc/xsltproc.c#L403) and that forwards that `FILE *` to [`xmlOutputBufferCreateFile(file, encoder)`](https://gitlab.gnome.org/GNOME/libxslt/-/blob/master/libxslt/xsltutils.c#L1722-1724), thus crossing each boundary (from xsltproc.exe's instance of libcmt.lib, into libxslt.dll, and then into libxml2.dll's use of api-ms-win-crt-stdio -> ucrtbase.dll).

**Review by SSE4 — APPROVED** [review]

**Pull request merged**